planner/core/hints.go

```diff
@@ -211,22 +216,53 @@ func getTableName(tblName model.CIStr, asName *model.CIStr) model.CIStr {
 	return tblName
 }
 
-func getJoinHints(joinType string, nodeType nodeType, children ...PhysicalPlan) (res []*ast.TableOptimizerHint) {
-	for i, child := range children {
-		table := extractTableAlias(child)
-		if table == nil {
+func extractTableAsName(p PhysicalPlan) *model.CIStr {
+	if len(p.Children()) > 1 {
+		return nil
+	}
+	switch x := p.(type) {
+	case *PhysicalTableReader:
+		ts := x.TablePlans[0].(*PhysicalTableScan)
+		if ts.TableAsName.L != "" {
+			return ts.TableAsName
+		}
+		return &ts.Table.Name
+	case *PhysicalIndexReader:
+		is := x.IndexPlans[0].(*PhysicalIndexScan)
+		if is.TableAsName.L != "" {
+			return is.TableAsName
+		}
+		return &is.Table.Name
+	case *PhysicalIndexLookUpReader:
+		is := x.IndexPlans[0].(*PhysicalIndexScan)
+		if is.TableAsName.L != "" {
+			return is.TableAsName
+		}
+		return &is.Table.Name
+	}
+	return nil
+}
+
+func getJoinHints(sctx sessionctx.Context, joinType string, parentOffset int, nodeType nodeType, children ...PhysicalPlan) (res []*ast.TableOptimizerHint) {
+	for _, child := range children {
+		if child.SelectBlockOffset() == -1 {
 			continue
 		}
-		// Merge the join hints with previous one if their select offsets are the same.
-		if len(res) > 0 && children[i].SelectBlockOffset() == children[i-1].SelectBlockOffset() {
-			res[len(res)-1].Tables = append(res[len(res)-1].Tables, ast.HintTable{TableName: table.name})
+		var tableName *model.CIStr
+		if child.SelectBlockOffset() != parentOffset {
+			tableName = &sctx.GetSessionVars().PlannerSelectBlockAsName[child.SelectBlockOffset()]
```

Comment:
Any test of this case?
Why the field name is `PlannerSelectBlockAsName`? Don't we need a table name here?